### PR TITLE
Implement regression

### PR DESCRIPTION
**Pull request by mlemainque**
This PR includes:

- A class `LGBMRegressor`
- New unit tests and system tests
- Avoid new sparse release because of broken unit tests (https://github.com/pydata/sparse/issues/218)

### CODE REVIEW

File: dask_lightgbm/core.py [hunk 1]
```diff
@@ -215,6 +214,13 @@ def predict_proba(self, X, client=None, **kwargs):
         return predict(client, self.to_local(), X, proba=True, dtype=self.classes_[0].dtype, **kwargs)
     predict_proba.__doc__ = lightgbm.LGBMClassifier.predict_proba.__doc__
 
+    def score(self, X, y, client=None, compute=True):
```
Comment (SfinxCZ) [inline dask_lightgbm/core.py]: I would rather not have this method here, the implementation of this wrapper should not be coupled with any metric. You can always call the code externally, if you need.

File: dask_lightgbm/core.py [hunk 2]
```diff
@@ -227,3 +233,60 @@ def to_local(self):
...
+            "machines": self.machines
+        }
+
+    def predict(self, X, client=None, **kwargs):
+        if client is None:
+            client = default_client()
+        return predict(client, self.to_local(), X, **kwargs)
+    predict.__doc__ = lightgbm.LGBMRegressor.predict.__doc__
+
+    def score(self, X, y, client=None, compute=True):
```
Comment (SfinxCZ) [inline dask_lightgbm/core.py]: Same comment as for the `LGBMClassifier` class - this method should not be here.

File: dask_lightgbm/core.py (hunk 1 above)
Comment (mlemainque) [reply dask_lightgbm/core.py]: I agree that the `.fit()` and `.score()` wrapper methods should not be coupled

In my training scripts I always print train & test scores but these computation do not fit into a single machine anymore... I though it would be comfortable to be able to migrate from local lightgbm to dask lightgbm without any modification of the existing code and use the same `model.score(X, y)` method

What about having an other class in an other package with the `.score()` implemented?

Comment (SfinxCZ) [reply dask_lightgbm/core.py]: Is there a reason why you cannot use http://ml.dask.org/modules/generated/dask_ml.metrics.accuracy_score.html#dask_ml.metrics.accuracy_score ? Because, `model.predict(X)` will return pointer to distributed collection rather than local numpy array. Therefore, if you use the `accuracy_score` defined in dask_ml it still will be distributed.

Mine line of thinking is that we should keep the API as similar to scikit as possible, and thus scoring functions should be decoupled from the model itself.

Comment (SfinxCZ) [reply dask_lightgbm/core.py]: Why you cannot use the `dask_ml.metrics.accuracy_score` directly? Because this code is mostly copy-paste from the dask-ml package. 
Quite frankly, I am not comfortable with shaping the library API for one particular use case and thus diverging from common API (scikit), especially when the functionality is already implemented in other library.

File: dask_lightgbm/core.py [hunk 3]
```diff
@@ -32,12 +32,12 @@ def parse_host_port(address):
...
 
 
-def build_network_params(worker_addresses, local_worker_ip, local_listen_port, listen_time_out):
+def build_network_params(worker_addresses, local_worker_ip, local_listen_port, time_out):
     addr_port_map = {addr: (local_listen_port + i) for i, addr in enumerate(worker_addresses)}
     params = {
         "machines": ",".join([parse_host_port(addr)[0] + ":" + str(port) for addr, port in addr_port_map.items()]),
         "local_listen_port": addr_port_map[local_worker_ip],
-        "listen_time_out": listen_time_out,
+        "time_out": time_out,
```
Comment (SfinxCZ) [inline dask_lightgbm/core.py]: I am not sure that this change is correct, I know that in the documentation is the configuration field called `time_out` but in the code it is `listen_time_out`, check https://github.com/Microsoft/LightGBM/blob/master/python-package/lightgbm/basic.py#L1529

File: system_tests/test_fit_predict.py [hunk 4]
```diff
@@ -20,18 +21,45 @@ def load_data(path):
 class FitPredictTest(unittest.TestCase):
```
Comment (SfinxCZ) [inline system_tests/test_fit_predict.py]: Please use `pytest` convention rather than `unittest` in order to keep the same code-style across the codebase.

File: system_tests/test_fit_predict.py [hunk 5]
```diff
@@ -1,37 +1,48 @@
...
-from dask import delayed
 from dask.distributed import Client
-from sklearn.metrics import confusion_matrix
 
 import dask_lightgbm.core as dlgbm
 
+if os.getenv('SCHEDULER'):
+    client = Client(os.getenv('SCHEDULER'))
+else:
+    client = Client()
```
Comment (SfinxCZ) [inline system_tests/test_fit_predict.py]: This could use the fixtures from pytest, so something like this:
```python
@pytest.yield_fixture(scope='module')
def client():
    with Client(os.getenv('SCHEDULER')) as client:
      yield client
```
and then all test that require client will have it as parameter, e.g.:

```python
def test_classify_newsread(client):
... rest of the code goes here...
```

**Comment by SfinxCZ** [comment]
@mlemainque I've tried to investigate these errors few weeks back and the result is that there is some problem with closing the communication in the LightGBM C++ code. Unfortunately, I was not able to precisely locate where is the problem. You may try to create issue for LightGBM directly. 

One solution (or more like hack) on our side is to wait for few seconds/minutes until the socket is timeouted and returned to OS or use different ports for every test (that is what I've used) or run every single test in separate docker container (ugh) since separate processes wont work.

File: system_tests/test_fit_predict.py [hunk 6]
```diff
@@ -1,37 +1,49 @@
...
 
-        print(confusion_matrix(dy.compute(), dy_pred.compute()))
 
-        self.assertGreaterEqual((dy == dy_pred).sum()/len(dy), 0.9)
+def test_regress_newsread(client):
+    data = dd.read_csv("./system_tests/data/*.gz", compression="gzip", blocksize=None)
+    dX = data.iloc[:, 1:]
+    dy = data.iloc[:, 0]
+
+    d_regress = dlgbm.LGBMRegressor(n_estimators=50, local_listen_port=13400)
```
[the remark this answers is not in the document]
Comment (SfinxCZ) [reply system_tests/test_fit_predict.py]: Maybe it would be much better to create fixture for port which would provide unique listen port so this would not happen in the future.

[the remark this answers is not in the document]
Comment (mlemainque) [reply system_tests/test_fit_predict.py]:
```python
@pytest.fixture()
def listen_port():
    listen_port.port += 10
    return listen_port.port

listen_port.port = 13000
```

I tried to use `itertools.count(13000, 10)` but no way to make it works inside a fixture

[the remark this answers is not in the document]
Comment (SfinxCZ) [reply system_tests/test_fit_predict.py]: Actually, this implementation seems ok to me.
